cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ArtifactHttpHandler.java

```diff
@@ -201,6 +210,172 @@ public void getArtifactInfo(HttpRequest request, HttpResponder responder,
+                              @PathParam("artifact-version") String artifactVersion)
+    throws NamespaceNotFoundException, BadRequestException {
+
+    Id.Namespace namespace = Id.Namespace.SYSTEM.getId().equalsIgnoreCase(namespaceId) ?
+      Id.Namespace.SYSTEM : validateAndGetNamespace(namespaceId);
+    Id.Artifact artifactId = validateAndGetArtifactId(namespace, artifactName, artifactVersion);
+
+    Map<String, String> properties;
+    try (Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8)) {
+      properties = GSON.fromJson(reader, MAP_STRING_STRING_TYPE);
+    } catch (JsonSyntaxException e) {
+      throw new BadRequestException("Json Syntax Error while parsing properties from request. " +
+                                      "Please check that the properties are a json map from string to string.", e);
+    } catch (IOException e) {
+      throw new BadRequestException("Unable to read properties from the request.", e);
+    }
+
+    try {
+      artifactRepository.writeArtifactProperties(artifactId, properties);
+      responder.sendStatus(HttpResponseStatus.OK);
+    } catch (ArtifactNotFoundException e) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, "Artifact " + artifactId + " not found.");
+    } catch (IOException e) {
+      LOG.error("Exception updating properties for artifact {}.", artifactId, e);
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Error adding properties to artifact.");
+    }
+  }
+
+  @PUT
+  @Path("/namespaces/{namespace-id}/artifacts/{artifact-name}/versions/{artifact-version}/properties/{property}")
+  public void writeProperty(HttpRequest request, HttpResponder responder,
+                            @PathParam("namespace-id") String namespaceId,
+                            @PathParam("artifact-name") String artifactName,
+                            @PathParam("artifact-version") String artifactVersion,
+                            @PathParam("property") String key)
+    throws NamespaceNotFoundException, BadRequestException {
+
+    Id.Namespace namespace = Id.Namespace.SYSTEM.getId().equalsIgnoreCase(namespaceId) ?
+      Id.Namespace.SYSTEM : validateAndGetNamespace(namespaceId);
+    Id.Artifact artifactId = validateAndGetArtifactId(namespace, artifactName, artifactVersion);
+
+    String value = request.getContent().toString(Charsets.UTF_8);
+
+    try {
+      artifactRepository.writeArtifactProperty(artifactId, key, value);
+      responder.sendStatus(HttpResponseStatus.OK);
+    } catch (ArtifactNotFoundException e) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, "Artifact " + artifactId + " not found.");
+    } catch (IOException e) {
+      LOG.error("Exception updating properties for artifact {}.", artifactId, e);
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Error writing property to artifact.");
```

Comment:
there is duplicate code between this method and ``writeProperties``, we can refactor and extract them into a common method..